src/unity/python/turicreate/toolkits/drawing_recognition/_sframe_loader.py

```diff
@@ -0,0 +1,282 @@
+        # thread. Each "message" is itself a _Queue that the background thread
+        # will use to communicate with us.
+        buffer_reset_queue = _Queue()
+        def worker():
+            data_source = _SFrameDataSource(
+                sframe, feature_column, target_column,
+                load_labels=load_labels, shuffle=shuffle, samples=samples)
+            while True:
+                buffer = buffer_reset_queue.get()
+                if buffer is None:
+                    break  # No more work to do, exit this thread.
+
+                for row in data_source:
+                    buffer.put(row)
+
+                    # Check if we've been reset (or told to exit).
+                    if not buffer_reset_queue.empty():
+                        break
+
+                # Always end each output buffer with None to signal completion.
+                buffer.put(None)
+                data_source.reset()
+        self.worker_thread = _Thread(target=worker)
+        self.worker_thread.daemon = True
+        self.worker_thread.start()
+
+        self.buffer_reset_queue = buffer_reset_queue
+        self.buffer_size = buffer_size
+
+        # Create the initial buffer and send it to the background thread, so
+        # that it begins sending us annotated images.
+        self.buffer = _Queue(self.buffer_size)
+        self.buffer_reset_queue.put(self.buffer)
+
+    def __del__(self):
+        # Tell the background thread to shut down.
+        self.buffer_reset_queue.put(None)
+
+        # Drain self.buffer to ensure that the background thread isn't stuck
+        # waiting to put something into it (and therefore never receives the
+        # shutdown request).
+        if self.buffer is not None:
+            while self.buffer.get() is not None:
+                pass
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return self._next()
+
+    def next(self):
+        return self._next()
+
+    def _next(self):
+        # Guard against repeated calls after we've finished.
+        if self.buffer is None:
+            raise StopIteration
+
+        result = self.buffer.get()
+        if result is None:
+            # Any future attempt to get from self.buffer will block forever,
+            # since the background thread won't put anything else into it.
+            self.buffer = None
+            raise StopIteration
+
+        return result
+
+    def reset(self):
+        # Send a new buffer to the background thread, telling it to reset.
+        buffer = _Queue(self.buffer_size)
+        self.buffer_reset_queue.put(buffer)
+
+        # Drain self.buffer to ensure that the background thread isn't stuck
+        # waiting to put something into it (and therefore never receives the
+        # new buffer).
+        if self.buffer is not None:
+            while self.buffer.get() is not None:
+                pass
+
+        self.buffer = buffer
+
+
+class SFrameRecognitionIter(_mx.io.DataIter):
```

Comment:
Why is this called a "recognition" iterator?

Reply:
To be in line with "drawing_recognition". As part of the "recognition" to "classifier" rename, I expect to rename this as well. Is there a better name you have in mind for this iterator?